src/internal/client.ts

```diff
@@ -911,6 +912,140 @@ export class TypedClient {
+    }
+    if (!isNumber(length)) {
+      throw new TypeError('length should be of type "number"')
+    }
+
+    let range = ''
+    if (offset || length) {
+      if (offset) {
+        range = `bytes=${+offset}-`
+      } else {
+        range = 'bytes=0-'
+        offset = 0
+      }
+      if (length) {
+        range += `${+length + offset - 1}`
+      }
+    }
+
+    const headers: RequestHeaders = {}
+    if (range !== '') {
+      headers.range = range
+    }
+
+    const expectedStatusCodes = [200]
+    if (range) {
+      expectedStatusCodes.push(206)
+    }
+    const method = 'GET'
+
+    const query = qs.stringify(getOpts)
+    return await this.makeRequestAsync({ method, bucketName, objectName, headers, query }, '', expectedStatusCodes)
+  }
+
+  /**
+   * download object content to a file.
+   * This method will create a temp file named `${filename}.${etag}.part.minio` when downloading.
+   *
+   * @param bucketName - name of the bucket
+   * @param objectName - name of the object
+   * @param filePath - path to which the object data will be written to
+   * @param getOpts - Optional object get option
+   */
+  async fGetObject(bucketName: string, objectName: string, filePath: string, getOpts: VersionIdentificator = {}) {
+    // Input validation.
+    if (!isValidBucketName(bucketName)) {
+      throw new errors.InvalidBucketNameError('Invalid bucket name: ' + bucketName)
+    }
+    if (!isValidObjectName(objectName)) {
+      throw new errors.InvalidObjectNameError(`Invalid object name: ${objectName}`)
+    }
+    if (!isString(filePath)) {
+      throw new TypeError('filePath should be of type "string"')
+    }
+
+    const downloadToTmpFile = async (): Promise<string> => {
+      let partFileStream: stream.Writable
+      const objStat = await this.statObject(bucketName, objectName, getOpts)
+      const partFile = `${filePath}.${objStat.etag}.part.minio`
+
+      await fsp.mkdir(path.dirname(filePath), { recursive: true })
+
+      let offset = 0
+      try {
+        const stats = await fsp.stat(partFile)
+        if (objStat.size === stats.size) {
+          return partFile
+        }
+        offset = stats.size
+        partFileStream = fs.createWriteStream(partFile, { flags: 'a' })
+      } catch (e) {
+        if (e instanceof Error && (e as unknown as { code: string }).code === 'ENOENT') {
+          // file not exist
+          partFileStream = fs.createWriteStream(partFile, { flags: 'w' })
+        } else {
+          // other error, maybe access deny
+          throw e
+        }
+      }
+
+      const downloadStream = await this.getPartialObject(bucketName, objectName, offset, 0, getOpts)
+
+      await streamPromise.pipeline(downloadStream, partFileStream)
+      const stats = await fsp.stat(partFile)
+      if (stats.size === objStat.size) {
+        return partFile
+      }
+
+      throw new Error('Size mismatch between downloaded file and the object')
```

Comment:
We should use a custom error instead of just regular `Error` type, it would be good if it also provides mismatched value, like:

```ts
class SizeMismatchError extends Error {
  constructor(public readonly objectSize: number, public readonly statSize: number) {
    super('Size mismatch between downloaded file and the object')
    this.name = 'SizeMismatchError'
  }
}